File: stepik-union/src/main/java/org/stepik/plugin/auth/webkit/network/ExtendedTime.java
package org.stepik.plugin.auth.webkit.network;

import org.jetbrains.annotations.NotNull;

/**
 * An extended time consisting of a long "base time" and
 * an integer "subtime".
 */
final class ExtendedTime implements Comparable<ExtendedTime> {
    private final long baseTime;
    private final int subtime;

    /**
     * Creates a new {@code ExtendedTime}.
     */
    private ExtendedTime(long baseTime, int subtime) {
        this.baseTime = baseTime;
        this.subtime = subtime;
    }

    /**
     * Returns the current extended time with the base time initialized
     * to System.currentTimeMillis() and the subtime initialized to zero.
     */
    @NotNull
    static ExtendedTime currentTime() {
        return new ExtendedTime(System.currentTimeMillis(), 0);
    }

    /**
     * Returns the base time.
     */
    long baseTime() {
        return baseTime;
    }

    /**
     * Returns the subtime.
     */
    int subtime() {
        return subtime;
    }

    /**
     * Increments the subtime and returns the result as a new extended time.
     */
    @NotNull
    ExtendedTime incrementSubtime() {
        return new ExtendedTime(baseTime, subtime + 1);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public int compareTo(@NotNull ExtendedTime otherExtendedTime) {
        int d = (int) (baseTime - otherExtendedTime.baseTime);
        if (d != 0) {
            return d;
        }
        return subtime - otherExtendedTime.subtime;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public String toString() {
        return "[baseTime=" + baseTime + ", subtime=" + subtime + "]";
    }
}

File: stepik-union/src/main/kotlin/org/stepik/core/auth/webkit/network/ExtendedTime.kt
package org.stepik.core.auth.webkit.network

/**
 * An extended time consisting of a long "base time" and
 * an integer "subtime".
 */
internal class ExtendedTime
/**
 * Creates a new `ExtendedTime`.
 */
private constructor(private val baseTime: Long, private val subtime: Int) : Comparable<ExtendedTime> {

    /**
     * Returns the base time.
     */
    fun baseTime(): Long {
        return baseTime
    }

    /**
     * Returns the subtime.
     */
    fun subtime(): Int {
        return subtime
    }

    /**
     * Increments the subtime and returns the result as a new extended time.
     */
    fun incrementSubtime(): ExtendedTime {
        return ExtendedTime(baseTime, subtime + 1)
    }

    /**
     * {@inheritDoc}
     */
    override fun compareTo(other: ExtendedTime): Int {
        val d = (baseTime - other.baseTime).toInt()
        return if (d != 0) {
            d
        } else subtime - other.subtime
    }

    /**
     * {@inheritDoc}
     */
    override fun toString(): String {
        return "[baseTime=$baseTime, subtime=$subtime]"
    }

    companion object {

        /**
         * Returns the current extended time with the base time initialized
         * to System.currentTimeMillis() and the subtime initialized to zero.
         */
        fun currentTime(): ExtendedTime {
            return ExtendedTime(System.currentTimeMillis(), 0)
        }
    }
}
